caas/kubernetes/provider/customresourcedefinitions.go

```diff
@@ -82,3 +107,224 @@ func (k *kubernetesClient) deleteCustomResourceDefinitions(appName string) error
+		crd, ok := crds[crdName]
+		if !ok {
+			// This should not happen.
+			return cleanUps, errors.NotFoundf("custom resource definition %q", crdName)
+		}
+		for _, crSpec := range crSpecList {
+			crdClient, err := k.getCustomResourceDefinitionClient(crd, getCRVersion(&crSpec))
+			if err != nil {
+				return cleanUps, errors.Trace(err)
+			}
+			crSpec.SetLabels(k.getCRLabels(appName))
+			_, crCleanUps, err := ensureCustomResource(crdClient, &crSpec)
+			cleanUps = append(cleanUps, crCleanUps...)
+			if err != nil {
+				return cleanUps, errors.Annotate(err, fmt.Sprintf("ensuring custom resource %q", crSpec.GetName()))
+			}
+			logger.Debugf("ensured custom resource %q", crSpec.GetName())
+		}
+	}
+	return cleanUps, nil
+}
+
+func ensureCustomResource(api dynamic.ResourceInterface, cr *unstructured.Unstructured) (out *unstructured.Unstructured, cleanUps []func(), err error) {
+	logger.Debugf("creating custom resource %q", cr.GetName())
+	if out, err = api.Create(cr); err == nil {
+		cleanUps = append(cleanUps, func() {
+			deleteCustomResourceDefinition(api, out.GetName(), out.GetUID())
+		})
+		return out, cleanUps, nil
+	}
+	if !k8serrors.IsAlreadyExists(err) {
+		return nil, cleanUps, errors.Trace(err)
+	}
+	// K8s complains about metadata.resourceVersion is required for an update, so get it before updating.
+	existingCR, err := api.Get(cr.GetName(), v1.GetOptions{})
+	if err != nil {
+		return nil, cleanUps, errors.Trace(err)
+	}
+	cr.SetResourceVersion(existingCR.GetResourceVersion())
+	logger.Debugf("updating custom resource %q", cr.GetName())
+	// TODO(caas): do label check to ensure the resource to be updated was created by Juju once caas upgrade steps of 2.7 in place.
+	out, err = api.Update(cr)
+	return out, cleanUps, errors.Trace(err)
+}
+
+func deleteCustomResourceDefinition(api dynamic.ResourceInterface, name string, uid types.UID) error {
+	err := api.Delete(name, newPreconditionDeleteOptions(uid))
+	if k8serrors.IsNotFound(err) {
+		return nil
+	}
+	return errors.Trace(err)
+}
+
+func (k *kubernetesClient) getCRDsForCRs(crs map[string][]unstructured.Unstructured) (out map[string]*apiextensionsv1beta1.CustomResourceDefinition, err error) {
+	var wg sync.WaitGroup
+	wg.Add(len(crs))
+	out = make(map[string]*apiextensionsv1beta1.CustomResourceDefinition)
+	crdChan := make(chan *apiextensionsv1beta1.CustomResourceDefinition, 1)
+	errChan := make(chan error, 1)
+
+	getCRD := func(
+		name string,
+		broker *kubernetesClient,
+		resultChan chan<- *apiextensionsv1beta1.CustomResourceDefinition,
+		errChan chan<- error,
+		clk jujuclock.Clock,
+	) {
+		checker := func() error {
+			crd, err := broker.getCustomResourceDefinition(name, false)
+			if err != nil {
+				return errors.Annotatef(err, "getting custom resource definition %q", name)
+			}
+			version := crd.Spec.Version
+			if version == "" {
+				if len(crd.Spec.Versions) == 0 {
+					return errors.NotValidf("custom resource definition %q no version", crd.GetName())
```

Comment:
s/no version/without a version

do we just want to log a warning and skip instead of returning an error

Reply:
no, version is `required` to generate the crd client.